compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java

```diff
@@ -1107,6 +1168,145 @@ public void visit(BLangSimpleVariable varNode) {
+        BLangType typeNode;
+        for (BLangSimpleVariable field : recordTypeNode.fields) {
+            typeNode = field.typeNode;
+            if (typeNode != null) {
+                typeNode.accept(this);
+            }
+        }
+        typeNode = recordTypeNode.restFieldType;
+        if (typeNode != null) {
+            typeNode.accept(this);
+        }
+    }
+
+    @Override
+    public void visit(BLangArrayType arrayType) {
+        arrayType.elemtype.accept(this);
+    }
+
+    @Override
+    public void visit(BLangConstrainedType constrainedType) {
+        constrainedType.constraint.accept(this);
+    }
+
+    @Override
+    public void visit(BLangErrorType errorType) {
+        if (errorType.detailType != null) {
+            errorType.detailType.accept(this);
+        }
+    }
+
+    @Override
+    public void visit(BLangFunctionTypeNode functionTypeNode) {
+        BLangType typeNode = functionTypeNode.returnTypeNode;
+        if (typeNode != null) {
+            typeNode.accept(this);
+        }
+        for (BLangSimpleVariable parameter : functionTypeNode.params) {
+            typeNode = parameter.typeNode;
+            if (typeNode != null) {
+                typeNode.accept(this);
+            }
+        }
+        BLangVariable restParam = functionTypeNode.restParam;
+        if (restParam != null) {
+            typeNode = restParam.typeNode;
+            if (typeNode != null) {
+                typeNode.accept(this);
+            }
+        }
+    }
+
+    @Override
+    public void visit(BLangBuiltInRefTypeNode builtInRefTypeNode) {
+    }
+
+    @Override
+    public void visit(BLangTableTypeNode tableTypeNode) {
+        tableTypeNode.constraint.accept(this);
+        BLangTableKeyTypeConstraint tableKeyTypeConstraint = tableTypeNode.tableKeyTypeConstraint;
+        if (tableKeyTypeConstraint != null) {
+            tableKeyTypeConstraint.accept(this);
+        }
+    }
+
+    @Override
+    public void visit(BLangStreamType streamType) {
+        streamType.type.accept(this);
+        streamType.constraint.accept(this);
+        BLangType error = streamType.error;
+        if (error != null) {
+            error.accept(this);
+        }
+    }
+
+    @Override
+    public void visit(BLangTableKeyTypeConstraint keyTypeConstraint) {
+        keyTypeConstraint.keyType.accept(this);
+    }
+
+    @Override
+    public void visit(BLangObjectTypeNode objectTypeNode) {
+        for (BLangSimpleVariable field : objectTypeNode.fields) {
+            BLangType typeNode = field.typeNode;
+            if (typeNode != null) {
+                typeNode.accept(this);
+            }
+        }
+    }
+
+    @Override
+    public void visit(BLangFiniteTypeNode finiteTypeNode) {
+    }
+
+    @Override
+    public void visit(BLangIntersectionTypeNode intersectionTypeNode) {
+        BType type = intersectionTypeNode.getBType();
+        if (type.tag != TypeTags.INTERSECTION) {
```

Comment:
When is this false?

Reply:
This happens when the type becomes a type reference. Remove this logic as part of the new changes.